tests/ui/uninit_vec.rs

```diff
@@ -0,0 +1,70 @@
+#![warn(clippy::uninit_vec)]
+
+use std::mem::MaybeUninit;
+
+#[derive(Default)]
+struct MyVec {
+    vec: Vec<u8>,
+}
+
+fn main() {
```

Comment:
Could you also add tests for `Vec::new()`, `Vec::default()` and `let vec: Vec<u8> = Default::default();` as they should also be covered by `get_vec_init_kind`.

I would also be interested if the `allow` attribute, works on the `set_len()` statement. Could you also add a test like this:

```rs
let vec: Vec<u8> = Vec::new();

// This should surpress only this lint (The user know that they are dooing somehting wrong xD)
#[allow(clippy::uninit_vec)]
unsafe {
    vec.set_len(1_000_000);
}
```

The test are good otherwise :upside_down_face:

Reply:
A separate lint would be needed for `Vec::new()` and `Vec::default()` cases since they do not precisely fall into "uninitialized elements" description. They create elements not only uninitialized but also out-of-bound, and the suggested fix should be different.

I think not linting against them is a viable option. People write code with `with_capacity()` and `set_len()` because they think such code is fine. On the other hand, `new()` or `default()` to `set_len()` are incorrect in more obvious ways, and I haven't seen a person write such code in managed codebase. A simple sanity test will catch them with segfault, so a dedicated lint might not be necessary.

I'm slightly inclined towards not linting against them without strong preference. I don't think many people are writing such code to dedicate a lint for such code pattern, but the lint would also require only a small amount of implementation effort. What's your thought?

Reply:
I think those cases should also be linted by either this or an additional lint. It might not be as obvious for some people that `Vec::new` and then `set_len()` allows access to out-of-bounds and uninitialized memory. Newcomers could for instance mix up `reserve` and `set_len`. The current implementation uses `get_vec_init_kind(cx, init_expr).is_some()` which will catch these cases anyway. Catching more mistakes if it's this simple is better than less, IMO.

In regard to splitting the lint, the name `uninit_vec` is general enough to also cover this case. Splitting the logic into two lints would enable the user to allow one and deny the other (We have split some lints for this reason). However, I see no reason why anyone would make this distinction. Therefore, I think that it's good as it is :upside_down_face: